Implement @supports with CSS nesting

mdubet opened the pull request:
#### c4adae4df0f33dcecf1ebcfa82bf5b37e2026ecf
<pre>
Implement @supports with CSS nesting
<a href="https://bugs.webkit.org/show_bug.cgi?id=252301">https://bugs.webkit.org/show_bug.cgi?id=252301</a>
rdar://105485711

Reviewed by Brent Fulgham.

The @supports(selector(...)) feature need to be aware of the
parser nesting context to properly works with relative nested selector.

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSupportsCondition):
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserEnum.h: Added
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeSupportsRule):
(WebCore::CSSParserImpl::consumeStyleRule):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::CSSSelectorParser::consumeNestedSelectorList):
(WebCore::CSSSelectorParser::supportsComplexSelector):
(WebCore::CSSSelectorParser::consumeNestedComplexSelector):
(WebCore::CSSSelectorParser::consumeSimpleSelector):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::supportsCondition):
(WebCore::CSSSupportsParser::consumeSupportsSelectorFunction):
* Source/WebCore/css/parser/CSSSupportsParser.h:
(WebCore::CSSSupportsParser::CSSSupportsParser):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::isValidRuleHeaderText):
(WebCore::isNestedContext):

Canonical link: <a href="https://commits.webkit.org/261004@main">https://commits.webkit.org/261004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6af2ee93a948948b0d3c492ce9c70289ce0563ef

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/110229 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/19326 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42885 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1648 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119217 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/114180 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20788 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10518 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/102496 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115973 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98684 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43700 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97437 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30348 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85548 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12015 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31685 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12627 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8650 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17993 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/51302 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7623 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14436 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/css/parser/CSSSupportsParser.h:
```diff
@@ -30,6 +30,7 @@
 #pragma once
 
 #include "CSSParserToken.h"
+#include "CSSSelectorParser.h"
```
You might consider making a CSSSelectorEnum header so we don't have to include the whole CSSSelectorParser (in line with @jernoble 's build improvement work!)

darinadler replied:
Sometimes you can just do a forward declaration and not need to create a new header.

brentfulgham commented on Source/WebCore/css/parser/CSSSelectorParser.cpp:
```diff
@@ -162,17 +157,22 @@ CSSSelectorList CSSSelectorParser::consumeForgivingComplexSelectorList(CSSParser
+    if (isNestedContext == IsNestedContext::Yes)
+        parserSelector = parser.consumeNestedComplexSelector(range);
+    else
+        parserSelector = parser.consumeComplexSelector(range);
```
Channeling @darinadler :

```
auto parserSelector = isNestedContext == IsNestedContext::Yes ?  parser.consumeNestedComplexSelector(range) : parser.consumeComplexSelector(range);
```

brentfulgham replied:
It does seem like `consumerNestedComplexSelector` always tries for `consumerComplexSelector` first, then tries the Relative case is that fails.

Maybe this would work:

```
auto parserSelector = parser.consumeComplexSelector(range);
if (!parserSelector && isNestedContext == IsNestedContext::Yes)
    parserSelector = consumeRelativeNestedSelector(range);
```

mdubet replied:
It does, but I don't think we get any benefit in writing it that way (it requires more knowledge of what is a nested selector for the caller for no reason).

brentfulgham commented on Source/WebCore/css/parser/CSSParserEnum.h:
```diff
@@ -0,0 +1,42 @@
+    Yes,
+    No
+};
+
```
We should eventually move `ParseResult` here, too.

brentfulgham reviewed (approved):
LGTM

mdubet committed:
Implement @supports with CSS nesting
https://bugs.webkit.org/show_bug.cgi?id=252301
rdar://105485711

Reviewed by Brent Fulgham.

The @supports(selector(...)) feature need to be aware of the
parser nesting context to properly works with relative nested selector.

* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule.html: Added.
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/parser/CSSParser.cpp:
(WebCore::CSSParser::parseSupportsCondition):
(WebCore::CSSParser::parseSelector):
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserEnum.h: Added
* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeSupportsRule):
(WebCore::CSSParserImpl::consumeStyleRule):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::parseCSSSelector):
(WebCore::CSSSelectorParser::CSSSelectorParser):
(WebCore::CSSSelectorParser::consumeNestedSelectorList):
(WebCore::CSSSelectorParser::supportsComplexSelector):
(WebCore::CSSSelectorParser::consumeNestedComplexSelector):
(WebCore::CSSSelectorParser::consumeSimpleSelector):
* Source/WebCore/css/parser/CSSSelectorParser.h:
* Source/WebCore/css/parser/CSSSupportsParser.cpp:
(WebCore::CSSSupportsParser::supportsCondition):
(WebCore::CSSSupportsParser::consumeSupportsSelectorFunction):
* Source/WebCore/css/parser/CSSSupportsParser.h:
(WebCore::CSSSupportsParser::CSSSupportsParser):
* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::isValidRuleHeaderText):
(WebCore::isNestedContext):

Canonical link: https://commits.webkit.org/261004@main
